Etcd log does not follow log format

In standalone mode, etcd log is using its own format like this:

```
[2020/02/20 22:06:06.122 +08:00] [WARN] [tidb_requests.go:71] ["get key failed"] [key=/tidb/server/info] [error="context deadline exceeded"]
[2020/02/20 22:06:54.254 +08:00] [ERROR] [periodic.go:53] ["can not get RegionsInfo"] [error="Get http://127.0.0.1:2379/pd/api/v1/regions: dial tcp 127.0.0.1:2379: connect: connection refused"] [stack="github.com/pingcap/log.Error\n\t/Users/breezewish/go/pkg/mod/github.com/pingcap/log@v0.0.0-20200117041106-d28c14d3b1cd/global.go:42\ngithub.com/pingcap-incubator/tidb-dashboard/pkg/keyvisual/input.(*periodicInput).Background\n\t/Users/breezewish/Work/PingCAP/tidb-dashboard/pkg/keyvisual/input/periodic.go:53\ngithub.com/pingcap-incubator/tidb-dashboard/pkg/keyvisual.(*Service).Start.func2\n\t/Users/breezewish/Work/PingCAP/tidb-dashboard/pkg/keyvisual/service.go:86"]
{"level":"warn","ts":"2020-02-20T22:06:55.248+0800","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"endpoint://client-72fa7705-b76a-47a4-a64e-8ff99c606a99/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest connection error: connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:2379: connect: connection refused\""}
```

which is annoying.


<!-- GanttProgress: 90% -->

--- a/pkg/pd/etcd.go
+++ b/pkg/pd/etcd.go
@@ -16,15 +16,32 @@ package pd
 import (
 	"time"
 
+	"github.com/pingcap/log"
 	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/keepalive"
 
 	"github.com/pingcap-incubator/tidb-dashboard/pkg/config"
 )
 
 const (
-	EtcdTimeout               = time.Second * 3
 	TiDBServerInformationPath = "/tidb/server/info"
 )
+
+func newZapEncoder(zapcore.EncoderConfig) (zapcore.Encoder, error) {
+	logCfg := log.Config{
+		Level: "info",
+		File:  log.FileLogConfig{},
+	}
+	return log.NewTextEncoder(&logCfg), nil
+}
+
+func init() {
+	_ = zap.RegisterEncoder("etcd-client", newZapEncoder)
+}
 
 var _ EtcdProvider = (*LocalEtcdProvider)(nil)
 
@@ -38,10 +55,36 @@ type LocalEtcdProvider struct {
 }
 
 func NewLocalEtcdClientProvider(config *config.Config) (*LocalEtcdProvider, error) {
+	// TODO: refactor
+	// Because etcd client does not support setting logger directly,
+	// the configuration of pingcap/log is copied here.
+	zapCfg := zap.NewProductionConfig()
+	zapCfg.Encoding = "etcd-client"
+	zapCfg.OutputPaths = []string{"stderr"}
+	zapCfg.ErrorOutputPaths = []string{"stderr"}
+
 	client, err := clientv3.New(clientv3.Config{
-		Endpoints:   []string{config.PDEndPoint},
-		DialTimeout: EtcdTimeout,
-		TLS:         config.TLSConfig,
+		Endpoints:        []string{config.PDEndPoint},
+		AutoSyncInterval: 30 * time.Second,
+		DialTimeout:      5 * time.Second,
+		DialOptions: []grpc.DialOption{
+			grpc.WithConnectParams(grpc.ConnectParams{
+				Backoff: backoff.Config{
+					BaseDelay:  1.0 * time.Second,
+					Multiplier: 1.6,
+					Jitter:     0.2,
+					MaxDelay:   3 * time.Second,
+				},
+				MinConnectTimeout: 20 * time.Second,
+			}),
+			grpc.WithKeepaliveParams(keepalive.ClientParameters{
+				Time:                10 * time.Second,
+				Timeout:             3 * time.Second,
+				PermitWithoutStream: true,
+			}),
+		},
+		TLS:       config.TLSConfig,
+		LogConfig: &zapCfg,
 	})
 	if err != nil {
 		return nil, err

--- a/pkg/pd/etcd.go
+++ b/pkg/pd/etcd.go
@@ -33,8 +33,8 @@ const (
 
 func newZapEncoder(zapcore.EncoderConfig) (zapcore.Encoder, error) {
 	logCfg := log.Config{
-		Level: "info",
-		File:  log.FileLogConfig{},
+		DisableTimestamp:    false,
+		DisableErrorVerbose: false,
 	}
 	return log.NewTextEncoder(&logCfg), nil
 }

